File: src/HPOE.java
/**
 * HPOE (Heuristic Page Optimization Engine) - Java Proposal
 * Includes Complete Regex Logic, Specs Objects, & Battery Saver Caps.
 */
import java.util.regex.*;
import java.util.*;

public class HPOE {
    public enum Tier { HIGH, MID, LOW, VERY_LOW }

    public static class HardwareSpecs {
        public String vendor = "Unknown";
        public String architecture = "Unknown";
        public String type = "Unknown";
        public String estimatedClass = "Unknown";
    }

    private Tier currentTier = Tier.HIGH;
    private HardwareSpecs specs = new HardwareSpecs();
    private boolean isMobileDevice = false;

    public HPOE() {
        evaluateSystemHardware();
        startPerformanceMonitor();
    }

    public void evaluateSystemHardware() {
        int coreCount = Runtime.getRuntime().availableProcessors();
        long memoryGB = Runtime.getRuntime().maxMemory() / (1024 * 1024 * 1024);
        int maxTextureSize = 8192; 

        String rawGpu = getMockGpuVendorString().toLowerCase();
        String renderer = rawGpu.replaceAll("\\(r|tm\\)", "").replaceAll("graphics", "").trim();
        Tier calculated = Tier.HIGH;

        // ---------------------------------------------------------
        // MASSIVE HEURISTIC GPU CLASSIFICATION MATRIX
        // ---------------------------------------------------------

        // 1. APPLE SILICON & A-SERIES
        if (renderer.contains("apple")) {
            specs.vendor = "Apple";
            if (Pattern.compile("m[1-9]").matcher(renderer).find()) {
                specs.architecture = "M-Series Silicon";
                specs.type = "Integrated";
                specs.estimatedClass = (renderer.contains("max") || renderer.contains("pro") || renderer.contains("ultra")) ? "High-End" : "Mid-Range";
                calculated = specs.estimatedClass.equals("High-End") ? Tier.HIGH : Tier.MID;
            } else {
                specs.architecture = "A-Series Silicon";
                specs.type = "Mobile";
                if (coreCount >= 6 && maxTextureSize >= 8192) {
                    specs.estimatedClass = "High-End";
                    calculated = Tier.HIGH;
                } else {
                    specs.estimatedClass = "Mid-Range";
                    calculated = Tier.MID;
                }
            }
        }
        // 2. NVIDIA DESKTOP & LAPTOP
        else if (renderer.contains("nvidia") || renderer.contains("geforce") || renderer.contains("quadro")) {
            specs.vendor = "NVIDIA";
            specs.type = "Discrete";
            if (Pattern.compile("rtx\\s*[2-9][0-9]{3}").matcher(renderer).find() || renderer.contains("titan") || Pattern.compile("quadro\\s*rtx").matcher(renderer).find()) {
                specs.architecture = "RTX / Ampere / Ada / Turing";
                specs.estimatedClass = "High-End";
                calculated = Tier.HIGH;
            } else if (Pattern.compile("gtx\\s*(?:10[6-9][0-9]|16[5-9][0-9]|9[8-9][0-9])").matcher(renderer).find()) {
                specs.architecture = "GTX High/Mid";
                specs.estimatedClass = "High-End";
                calculated = Tier.HIGH;
            } else if (Pattern.compile("gtx\\s*(?:1050|970|960|950|780)").matcher(renderer).find()) {
                specs.architecture = "GTX Legacy Mid";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier.MID;
            } else if (Pattern.compile("gtx\\s*(?:4|5|6)[0-9]{2}").matcher(renderer).find() || Pattern.compile("gt\\s*[0-9]+").matcher(renderer).find() || Pattern.compile("[7-9][1-4]0m").matcher(renderer).find() || Pattern.compile("mx[1-4][0-9]{2}").matcher(renderer).find()) {
                specs.architecture = "Legacy / Low-End Mobile (GT/MX)";
                specs.estimatedClass = "Budget/Legacy";
                calculated = Tier.LOW;
            } else {
                specs.architecture = "Unknown NVIDIA";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier.MID;
            }
        }
        // 3. AMD RADEON
        else if (renderer.contains("amd") || renderer.contains("radeon")) {
            specs.vendor = "AMD";
            if (Pattern.compile("rx\\s*[5-8][0-9]{3}").matcher(renderer).find() || Pattern.compile("vega\\s*(?:56|64)").matcher(renderer).find() || renderer.contains("radeon pro")) {
                specs.architecture = "RDNA / High-End Vega";
                specs.type = "Discrete";
                specs.estimatedClass = "High-End";
                calculated = Tier.HIGH;
            } else if (Pattern.compile("rx\\s*(?:4[0-9]{2}|5[7-9][0-9])").matcher(renderer).find() || Pattern.compile("6[6-9]0m").matcher(renderer).find()) {
                specs.architecture = "Polaris / Modern APU";
                specs.type = "Integrated";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier.MID;
            } else {
                specs.architecture = "Legacy GCN or Budget APU";
                specs.type = "Integrated";
                specs.estimatedClass = "Budget/Legacy";
                calculated = Tier.LOW;
            }
        }
        // 4. INTEL GRAPHICS
        else if (renderer.contains("intel")) {
            specs.vendor = "Intel";
            if (Pattern.compile("arc\\s*a[3-7][0-9]{2}").matcher(renderer).find()) {
                specs.architecture = "Arc Alchemist";
                specs.type = "Discrete";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier.MID;
            } else if ((renderer.contains("iris") && renderer.contains("xe")) || Pattern.compile("iris\\s*(?:plus|pro)").matcher(renderer).find()) {
                specs.architecture = "Iris Xe/Plus";
                specs.type = "Integrated";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier.MID;
            } else {
                specs.architecture = "UHD / HD Legacy";
                specs.type = "Integrated";
                specs.estimatedClass = "Budget/Legacy";
                calculated = Tier.LOW;
            }
        }
        // 5. QUALCOMM ADRENO / SNAPDRAGON (ANDROID)
        else if (renderer.contains("adreno") || renderer.contains("snapdragon")) {
            specs.vendor = "Qualcomm";
            specs.type = "Mobile";
            Matcher m = Pattern.compile("adreno\\s*([0-9]{3})").matcher(renderer);
            int series = (m.find()) ? Integer.parseInt(m.group(1)) : 0;
            if (series >= 800 || renderer.contains("snapdragon 8 elite") || renderer.contains("elite")) {
                specs.architecture = (series != 0) ? "Adreno " + series : "Snapdragon 8 Elite Flagship";
                specs.estimatedClass = "High-End";
                calculated = Tier.HIGH;
            } else if (series >= 730 || renderer.contains("snapdragon 8 gen")) {
                specs.architecture = (series != 0) ? "Adreno " + series : "Snapdragon 8 Gen Flagship";
                specs.estimatedClass = "High-End";
                calculated = Tier.HIGH;
            } else if (series >= 650 || renderer.contains("snapdragon 8") || renderer.contains("snapdragon 7")) {
                specs.architecture = (series != 0) ? "Adreno " + series : "Snapdragon 7/8 Series";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier.MID;
            } else if ((series == 0 && coreCount >= 8 && maxTextureSize >= 8192) || renderer.contains("snapdragon")) {
                specs.architecture = "Modern Adreno/Snapdragon (Masked)";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier.MID;
            } else {
                specs.architecture = "Adreno " + ((series != 0) ? series : "Legacy");
                specs.estimatedClass = "Budget/Legacy";
                calculated = Tier.LOW;
            }
        }
        // 6. ARM MALI (ANDROID)
        else if (renderer.contains("mali")) {
            specs.vendor = "ARM";
            specs.type = "Mobile";
            if (renderer.contains("immortalis") || Pattern.compile("g[7-9][1-9][0-9]").matcher(renderer).find() || Pattern.compile("g7[7-9]").matcher(renderer).find()) {
                specs.architecture = "Mali Valhall / Immortalis Flagship";
                specs.estimatedClass = "High-End";
                calculated = Tier.HIGH;
            } else if (Pattern.compile("g[7-9][0-9]").matcher(renderer).find()) {
                specs.architecture = "Mali Valhall / Immortalis";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier.MID;
            } else {
                specs.architecture = "Legacy Mali / Budget G-Series";
                specs.estimatedClass = "Budget/Legacy";
                calculated = Tier.LOW;
            }
        }
        // 7. POWERVR (OLDER IOS / MEDIATEK)
        else if (renderer.contains("powervr")) {
            specs.vendor = "Imagination Technologies";
            specs.type = "Mobile";
            specs.architecture = "PowerVR Rogue/SGX";
            specs.estimatedClass = "Budget/Legacy";
            calculated = Tier.LOW;
        }
        // 8. SAMSUNG XCLIPSE / EXYNOS (MOBILE)
        else if (renderer.contains("xclipse") || renderer.contains("exynos")) {
            specs.vendor = "Samsung";
            specs.type = "Mobile";
            Matcher m = Pattern.compile("xclipse\\s*([0-9]{3})").matcher(renderer);
            if (m.find()) {
                int series = Integer.parseInt(m.group(1));
                if (series >= 920) {
                    specs.architecture = "Xclipse " + series + " (RDNA Flagship)";
                    specs.estimatedClass = "High-End";
                    calculated = Tier.HIGH;
                } else {
                    specs.architecture = "Xclipse " + series + " (RDNA)";
                    specs.estimatedClass = "Mid-Range";
                    calculated = Tier.MID;
                }
            } else {
                if (coreCount >= 8 && maxTextureSize >= 8192) {
                    specs.architecture = "Exynos Flagship";
                    specs.estimatedClass = "High-End";
                    calculated = Tier.HIGH;
                } else {
                    specs.architecture = "Exynos Legacy/Masked";
                    specs.estimatedClass = "Budget/Legacy";
                    calculated = Tier.LOW;
                }
            }
        }
        // 9. MEDIATEK (DIMENSITY / HELIO)
        else if (renderer.contains("mediatek") || renderer.contains("dimensity") || renderer.contains("helio")) {
            specs.vendor = "MediaTek";
            specs.type = "Mobile";
            if (renderer.contains("dimensity 9") || renderer.contains("dimensity 8")) {
                specs.architecture = "Dimensity 8000/9000 Flagship";
                specs.estimatedClass = "High-End";
                calculated = Tier.HIGH;
            } else if (renderer.contains("dimensity") || renderer.contains("helio g9") || (coreCount >= 8 && maxTextureSize >= 8192)) {
                specs.architecture = "Dimensity / High-End Helio";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier.MID;
            } else {
                specs.architecture = "Helio Legacy / Entry SoC";
                specs.estimatedClass = "Budget/Legacy";
                calculated = Tier.LOW;
            }
        }
        // 10. UNISOC / SPREADTRUM
        else if (renderer.contains("unisoc") || renderer.contains("spreadtrum") || renderer.contains("tigert")) {
            specs.vendor = "Unisoc";
            specs.type = "Mobile";
            specs.architecture = "Unisoc / Spreadtrum";
            specs.estimatedClass = "Budget/Legacy";
            calculated = Tier.LOW;
        } 
        // 11. FALLBACK FOR UNKNOWN GPUs
        else {
            specs.estimatedClass = "Budget/Legacy";
            calculated = Tier.LOW;
            specs.architecture = "Unknown (Fallback)";
        }

        // Hard limits and Accessibility Overrides
        if (coreCount <= 2 && memoryGB <= 2) calculated = Tier.VERY_LOW;
        else if (coreCount <= 2 || memoryGB <= 2) calculated = Tier.LOW;
        else if (memoryGB <= 3 && calculated == Tier.HIGH) calculated = Tier.MID;
        
        // Absolute fail-safe: Mobile devices NEVER get "High" tier effects.
        if (isMobileDevice && calculated == Tier.HIGH) calculated = Tier.MID;

        currentTier = calculated;
    }

    // ---------------------------------------------------------
    // LIVE V-SYNC DEGRADATION MONITOR (FPS SAFETY NET)
    // ---------------------------------------------------------
    private void startPerformanceMonitor() {
        Thread t = new Thread(() -> {
            int sustainedDropTicks = 0;
            int detectedBaselineFps = 60;
            int batterySaverTicks = 0;
            int gracePeriodTicks = 0;
            List<Integer> baselineMeasurements = new ArrayList<>();

            while (currentTier != Tier.LOW && currentTier != Tier.VERY_LOW) {
                try {
                    Thread.sleep(1000);
                    int fps = getMockSystemFps();
                    int maxFrameDelta = getMockMaxFrameDelta();

                    gracePeriodTicks++;
                    if (gracePeriodTicks <= 3) {
                        if (gracePeriodTicks > 1 && fps > 0) baselineMeasurements.add(fps);
                        continue;
                    }

                    if (baselineMeasurements.size() > 0 && detectedBaselineFps == 60) {
                        double avg = baselineMeasurements.stream().mapToInt(val -> val).average().orElse(0.0);
                        if (avg >= 28 && avg <= 34 && maxFrameDelta < 45) {
                            detectedBaselineFps = 30;
                        }
                    }

                    if (detectedBaselineFps == 60 && fps >= 28 && fps <= 33 && maxFrameDelta < 45) {
                        batterySaverTicks++;
                        if (batterySaverTicks >= 2) {
                            detectedBaselineFps = 30;
                            sustainedDropTicks = 0;
                        }
                    } else if (detectedBaselineFps == 30 && fps >= 45) {
                        detectedBaselineFps = 60;
                        sustainedDropTicks = 0;
                        batterySaverTicks = 0;
                    } else if (detectedBaselineFps == 60 && (fps < 28 || fps > 33 || maxFrameDelta >= 45)) {
                        batterySaverTicks = Math.max(0, batterySaverTicks - 1);
                    }

                    int floor = (detectedBaselineFps == 30) ? ((currentTier == Tier.HIGH) ? 22 : 18) : ((currentTier == Tier.HIGH) ? 45 : 38);
                    int limit = (currentTier == Tier.HIGH) ? 4 : 6;

                    if (fps < floor) sustainedDropTicks++;
                    else sustainedDropTicks = Math.max(0, sustainedDropTicks - 2);

                    if (sustainedDropTicks >= limit) {
                        currentTier = (currentTier == Tier.HIGH) ? Tier.MID : Tier.LOW;
                        System.out.println("[Hardware Profiler] Emergency downgrading tier.");
                        sustainedDropTicks = 0;
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        });
        t.setDaemon(true);
        t.start();
    }

    private String getMockGpuVendorString() { return "Intel Iris Plus Graphics"; }
    private int getMockSystemFps() { return 60; }
    private int getMockMaxFrameDelta() { return 16; }
}

File: src/HPOE.cpp
#include <iostream>
#include <string>
#include <thread>
#include <chrono>
#include <algorithm>
#include <regex>
#include <vector>
#include <numeric>

/**
 * HPOE (Heuristic Page Optimization Engine) - C++ Proposal
 * Full Regex Matrix, Specs Objects & Battery Saver Trackers.
 */

enum class Tier { HIGH, MID, LOW, VERY_LOW };

struct HardwareSpecs {
    std::string vendor = "Unknown";
    std::string architecture = "Unknown";
    std::string type = "Unknown";
    std::string estimatedClass = "Unknown";
};

class HPOE {
private:
    Tier currentTier;
    HardwareSpecs specs;
    bool isMonitoring;
    bool isMobile;

    void evaluateHardware() {
        int coreCount = std::thread::hardware_concurrency();
        int memoryGB = 16; 
        int maxTextureSize = 8192; 
        
        std::string rawGpu = getMockGpuVendorString();
        std::transform(rawGpu.begin(), rawGpu.end(), rawGpu.begin(), ::tolower);
        
        std::string renderer = std::regex_replace(rawGpu, std::regex("\\(r\\)|\\(tm\\)|graphics"), "");
        Tier calculated = Tier::HIGH;

        auto test = [&](const std::string& pat) { return std::regex_search(renderer, std::regex(pat)); };

        // ---------------------------------------------------------
        // MASSIVE HEURISTIC GPU CLASSIFICATION MATRIX
        // ---------------------------------------------------------

        // 1. APPLE SILICON & A-SERIES
        if (renderer.find("apple") != std::string::npos) {
            specs.vendor = "Apple";
            if (test("m[1-9]")) {
                specs.architecture = "M-Series Silicon";
                specs.type = "Integrated";
                specs.estimatedClass = (renderer.find("max") != std::string::npos || renderer.find("pro") != std::string::npos || renderer.find("ultra") != std::string::npos) ? "High-End" : "Mid-Range";
                calculated = specs.estimatedClass == "High-End" ? Tier::HIGH : Tier::MID;
            } else {
                specs.architecture = "A-Series Silicon";
                specs.type = "Mobile";
                if (coreCount >= 6 && maxTextureSize >= 8192) {
                    specs.estimatedClass = "High-End";
                    calculated = Tier::HIGH;
                } else {
                    specs.estimatedClass = "Mid-Range";
                    calculated = Tier::MID;
                }
            }
        }
        // 2. NVIDIA DESKTOP & LAPTOP
        else if (renderer.find("nvidia") != std::string::npos || renderer.find("geforce") != std::string::npos || renderer.find("quadro") != std::string::npos) {
            specs.vendor = "NVIDIA";
            specs.type = "Discrete";
            if (test("rtx\\s*[2-9][0-9]{3}") || renderer.find("titan") != std::string::npos || test("quadro\\s*rtx")) {
                specs.architecture = "RTX / Ampere / Ada / Turing";
                specs.estimatedClass = "High-End";
                calculated = Tier::HIGH;
            } else if (test("gtx\\s*(?:10[6-9][0-9]|16[5-9][0-9]|9[8-9][0-9])")) {
                specs.architecture = "GTX High/Mid";
                specs.estimatedClass = "High-End";
                calculated = Tier::HIGH;
            } else if (test("gtx\\s*(?:1050|970|960|950|780)")) {
                specs.architecture = "GTX Legacy Mid";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier::MID;
            } else if (test("gtx\\s*(?:4|5|6)[0-9]{2}") || test("gt\\s*[0-9]+") || test("[7-9][1-4]0m") || test("mx[1-4][0-9]{2}")) {
                specs.architecture = "Legacy / Low-End Mobile (GT/MX)";
                specs.estimatedClass = "Budget/Legacy";
                calculated = Tier::LOW;
            } else {
                specs.architecture = "Unknown NVIDIA";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier::MID;
            }
        }
        // 3. AMD RADEON
        else if (renderer.find("amd") != std::string::npos || renderer.find("radeon") != std::string::npos) {
            specs.vendor = "AMD";
            if (test("rx\\s*[5-8][0-9]{3}") || test("vega\\s*(?:56|64)") || renderer.find("radeon pro") != std::string::npos) {
                specs.architecture = "RDNA / High-End Vega";
                specs.type = "Discrete";
                specs.estimatedClass = "High-End";
                calculated = Tier::HIGH;
            } else if (test("rx\\s*(?:4[0-9]{2}|5[7-9][0-9])") || test("6[6-9]0m")) {
                specs.architecture = "Polaris / Modern APU";
                specs.type = "Integrated";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier::MID;
            } else {
                specs.architecture = "Legacy GCN or Budget APU";
                specs.type = "Integrated";
                specs.estimatedClass = "Budget/Legacy";
                calculated = Tier::LOW;
            }
        }
        // 4. INTEL GRAPHICS
        else if (renderer.find("intel") != std::string::npos) {
            specs.vendor = "Intel";
            if (test("arc\\s*a[3-7][0-9]{2}")) {
                specs.architecture = "Arc Alchemist";
                specs.type = "Discrete";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier::MID;
            } else if ((renderer.find("iris") != std::string::npos && renderer.find("xe") != std::string::npos) || test("iris\\s*(?:plus|pro)")) {
                specs.architecture = "Iris Xe/Plus";
                specs.type = "Integrated";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier::MID;
            } else {
                specs.architecture = "UHD / HD Legacy";
                specs.type = "Integrated";
                specs.estimatedClass = "Budget/Legacy";
                calculated = Tier::LOW;
            }
        }
        // 5. QUALCOMM ADRENO / SNAPDRAGON (ANDROID)
        else if (renderer.find("adreno") != std::string::npos || renderer.find("snapdragon") != std::string::npos) {
            specs.vendor = "Qualcomm";
            specs.type = "Mobile";
            std::smatch match;
            int series = 0;
            if (std::regex_search(renderer, match, std::regex("adreno\\s*([0-9]{3})"))) series = std::stoi(match[1].str());
            
            if (series >= 800 || renderer.find("snapdragon 8 elite") != std::string::npos || renderer.find("elite") != std::string::npos) {
                specs.architecture = (series != 0) ? "Adreno " + std::to_string(series) : "Snapdragon 8 Elite Flagship";
                specs.estimatedClass = "High-End";
                calculated = Tier::HIGH;
            } else if (series >= 730 || renderer.find("snapdragon 8 gen") != std::string::npos) {
                specs.architecture = (series != 0) ? "Adreno " + std::to_string(series) : "Snapdragon 8 Gen Flagship";
                specs.estimatedClass = "High-End";
                calculated = Tier::HIGH;
            } else if (series >= 650 || renderer.find("snapdragon 8") != std::string::npos || renderer.find("snapdragon 7") != std::string::npos) {
                specs.architecture = (series != 0) ? "Adreno " + std::to_string(series) : "Snapdragon 7/8 Series";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier::MID;
            } else if ((series == 0 && coreCount >= 8 && maxTextureSize >= 8192) || renderer.find("snapdragon") != std::string::npos) {
                specs.architecture = "Modern Adreno/Snapdragon (Masked)";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier::MID;
            } else {
                specs.architecture = "Adreno " + ((series != 0) ? std::to_string(series) : "Legacy");
                specs.estimatedClass = "Budget/Legacy";
                calculated = Tier::LOW;
            }
        }
        // 6. ARM MALI (ANDROID)
        else if (renderer.find("mali") != std::string::npos) {
            specs.vendor = "ARM";
            specs.type = "Mobile";
            if (renderer.find("immortalis") != std::string::npos || test("g[7-9][1-9][0-9]") || test("g7[7-9]")) {
                specs.architecture = "Mali Valhall / Immortalis Flagship";
                specs.estimatedClass = "High-End";
                calculated = Tier::HIGH;
            } else if (test("g[7-9][0-9]")) {
                specs.architecture = "Mali Valhall / Immortalis";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier::MID;
            } else {
                specs.architecture = "Legacy Mali / Budget G-Series";
                specs.estimatedClass = "Budget/Legacy";
                calculated = Tier.LOW;
            }
        }
        // 7. SAMSUNG XCLIPSE / EXYNOS (MOBILE)
        else if (renderer.find("xclipse") != std::string::npos || renderer.find("exynos") != std::string::npos) {
            specs.vendor = "Samsung";
            specs.type = "Mobile";
            std::smatch match;
            if (std::regex_search(renderer, match, std::regex("xclipse\\s*([0-9]{3})"))) {
                int series = std::stoi(match[1].str());
                if (series >= 920) {
                    specs.architecture = "Xclipse " + std::to_string(series) + " (RDNA Flagship)";
                    specs.estimatedClass = "High-End";
                    calculated = Tier::HIGH;
                } else {
                    specs.architecture = "Xclipse " + std::to_string(series) + " (RDNA)";
                    specs.estimatedClass = "Mid-Range";
                    calculated = Tier::MID;
                }
            } else {
                if (coreCount >= 8 && maxTextureSize >= 8192) {
                    specs.architecture = "Exynos Flagship";
                    specs.estimatedClass = "High-End";
                    calculated = Tier::HIGH;
                } else {
                    specs.architecture = "Exynos Legacy/Masked";
                    specs.estimatedClass = "Budget/Legacy";
                    calculated = Tier::LOW;
                }
            }
        }
        // 8. MEDIATEK (DIMENSITY / HELIO)
        else if (renderer.find("mediatek") != std::string::npos || renderer.find("dimensity") != std::string::npos || renderer.find("helio") != std::string::npos) {
            specs.vendor = "MediaTek";
            specs.type = "Mobile";
            if (renderer.find("dimensity 9") != std::string::npos || renderer.find("dimensity 8") != std::string::npos) {
                specs.architecture = "Dimensity 8000/9000 Flagship";
                specs.estimatedClass = "High-End";
                calculated = Tier::HIGH;
            } else if (renderer.find("dimensity") != std::string::npos || renderer.find("helio g9") != std::string::npos || (coreCount >= 8 && maxTextureSize >= 8192)) {
                specs.architecture = "Dimensity / High-End Helio";
                specs.estimatedClass = "Mid-Range";
                calculated = Tier.MID;
            } else {
                specs.architecture = "Helio Legacy / Entry SoC";
                specs.estimatedClass = "Budget/Legacy";
                calculated = Tier.LOW;
            }
        }
        // 9. POWERVR / UNISOC / SPREADTRUM
        else if (renderer.find("powervr") != std::string::npos || renderer.find("unisoc") != std::string::npos || renderer.find("spreadtrum") != std::string::npos || renderer.find("tigert") != std::string::npos) {
            specs.vendor = "Unisoc / PowerVR";
            specs.type = "Mobile";
            specs.architecture = "Legacy Budget";
            specs.estimatedClass = "Budget/Legacy";
            calculated = Tier::LOW;
        } 
        // 10. FALLBACK FOR UNKNOWN GPUs
        else {
            specs.estimatedClass = "Budget/Legacy";
            calculated = Tier::LOW;
            specs.architecture = "Unknown (Fallback)";
        }

        // Hard limits and Accessibility Overrides
        if (coreCount < 3 && memoryGB < 3) calculated = Tier::VERY_LOW;
        
        else if (memoryGB <= 3 && calculated == Tier::HIGH) calculated = Tier::MID;
        
        // Absolute fail-safe
        if (isMobile && calculated == Tier::HIGH) calculated = Tier::MID;

        currentTier = calculated;
    }

    // ---------------------------------------------------------
    // LIVE V-SYNC DEGRADATION MONITOR (FPS SAFETY NET)
    // ---------------------------------------------------------
    void degradationLoop() {
        int sustainedDrops = 0;
        int detectedBaseline = 60;
        int batterySaverTicks = 0;
        int gracePeriodTicks = 0;
        bool isBaselineLocked = false;
        std::vector<int> measurements;

        while (isMonitoring && currentTier != Tier::VERY_LOW && currentTier != Tier::LOW) {
            std::this_thread::sleep_for(std::chrono::seconds(1));
            int fps = getMockSystemFps();
            int maxFrameDelta = getMockMaxFrameDelta();

            gracePeriodTicks++;
            if (gracePeriodTicks <= 3) {
                if (gracePeriodTicks > 1 && fps > 0) measurements.push_back(fps);
                continue;
            }

            if (!measurements.empty() && !isBaselineLocked) {
                double avg = std::accumulate(measurements.begin(), measurements.end(), 0.0) / measurements.size();
                if (avg > 65) {
                    detectedBaseline = (int)(avg + 0.5);
                } else if (avg >= 28 && avg <= 34 && maxFrameDelta < 45) {
                    detectedBaseline = 30;
                } else {
                    detectedBaseline = 60;
                }
                isBaselineLocked = true;
            }

            if (detectedBaseline >= 60 && fps >= 28 && fps <= 33 && maxFrameDelta < 45) {
                batterySaverTicks++;
                if (batterySaverTicks >= 2) {
                    detectedBaseline = 30;
                    sustainedDrops = 0; 
                }
            } else if (detectedBaseline == 30 && fps >= 45) {
                detectedBaseline = (fps > 65) ? fps : 60;
                sustainedDrops = 0;
                batterySaverTicks = 0;
            } else if (detectedBaseline >= 60 && (fps < 28 || fps > 33 || maxFrameDelta >= 45)) {
                batterySaverTicks = std::max(0, batterySaverTicks - 1);
            }

            int floor = (detectedBaseline == 30) ? 22 : 40;
            int limit = 3;

            if (fps < floor) sustainedDrops++;
            else sustainedDrops = std::max(0, sustainedDrops - 2);

            if (sustainedDrops >= limit) {
                currentTier = (currentTier == Tier::HIGH) ? Tier::MID : Tier::LOW;
                std::cout << "[Hardware Profiler] Throttling detected. Downgraded tier.\n";
                sustainedDrops = 0;
            }
        }
    }

    std::string getMockGpuVendorString() { return "NVIDIA GeForce RTX 4080"; }
    int getMockSystemFps() { return 60; }
    int getMockMaxFrameDelta() { return 16; }

public:
    HPOE(bool mobile = false) : isMobile(mobile), isMonitoring(true) {
        evaluateHardware();
        std::thread(&HPOE::degradationLoop, this).detach();
    }
    ~HPOE() { isMonitoring = false; }
};
